### ISSUE TITLE
`<-` support in PSCi

### ISSUE THREAD

**Issue by noraesae** [purescript/purescript#2628]
In GHCi, we can bind an `IO` result into a name with `<-`.

```
> a <- return "hello" :: IO String
> a
"hello"
> :t a
a :: String
```

In PSCi, we can surely use `unsafePerformEff`, but I think it may be handy to support `<-` too.

```
> let eff = pure "hello" :: forall e. Eff e String
> let a = unsafePerformEff eff
> b <- eff
```

I'm thinking about implementing this, but first I'd like to hear some ideas about this.

**Comment by hdgarrood** [comment]
I'd very much like to support this. I guess the obvious implementation would be to treat `x <- y` as `let x = unsafePerformEff y`; I can't immediately see any problems with doing it that way as long as you're careful about scoping and so on - we don't want a situation where `<-` only works if you've imported Control.Monad.Eff.Unsafe, for instance. Other than that I don't have anything to add in terms of implementation.

**Comment by garyb** [comment]
I have 2p to add: whatever we do, let's not do what `ghci` does where you're in some weird ambient monad that makes things work magically without making it clear what is going on.

**Comment by hdgarrood** [comment]
Are you able to expand on that? I'm not sure what you mean.

**Comment by garyb** [comment]
I mean things where things like this happen:

```
Prelude> pure True
True
```

So I guess it's not so much about `_ <- x`, but more a case of let's be explicit about when we're doing monadic things and don't just "auto-bind" when evaluating.

**Comment by noraesae** [comment]
I can agree in being explicit. I'm okay with not auto-binding `Eff` too, but there could be people who find it handy. I also think then we may need `<-` in advance if there's no auto-binding.

```
> pure true
Error found:
in module $PSCI

  No type class instance was found for

    Eff t0 Boolean

...

> a <- pure true
> a
true
```

**Comment by hdgarrood** [comment]
I would hope that `pure true` in psci fails with an ambiguous type variable error right now, is that not the case?

**Comment by noraesae** [comment]
Oh, actually it is.

```
> pure true
Error found:
in module $PSCI

  No type class instance was found for

    Data.Show.Show (t2 Boolean)

  The instance head contains unknown type variables. Consider adding a type annotation.

while checking that expression eval it
  has type Eff t0 t1
while applying a function eval
  of type forall a.
            (Eval a) => a
                        -> Eff
                             ( "console" :: CONSOLE
                             )
                             Unit
  to argument it
in value declaration $main

where t0 is an unknown type
      t1 is an unknown type
      t2 is an unknown type

See https://github.com/purescript/documentation/blob/master/errors/NoInstanceFound.md for more information,
or to contribute content related to this error.
```

Sorry for the confusion.

**Comment by noraesae** [comment]
Now I understand how GHCi works (weirdly) with `<-` or evaluation. It just auto-binds everything, even `pure True`, though it's actually not a monad.

In PSCi, `<-` or evaluation shall only bind `Eff e a`. It looks okay?

**Comment by hdgarrood** [comment]
Cool :) psci's current behaviour seems ok to me. I also think it would be okay for `<-` in psci to require the RHS to match `Eff _ _`, and to be able to use that in type inference. That is, I don't see a problem with having `a <- pure true` work in psci.

**Comment by paf31** [comment]
It's not as simple as just using `unsafePerformEff`. Let bindings get reevaluated on each command in PSCi, which means monadic actions would get re-run. I have a plan for this, which first involves moving evaluation into a long-running Node process, like we do with the browser backend. Then, we need some sort of store of values on the Node side, which can hold onto values computed by `lets` and binds.

Finally, just supporting `Eff` is not enough. We should use a type class in `psci-support` to allow other non-core libraries to support this sort of evaluation in PSCi, like we have the `Eval` class now.

**Comment by hdgarrood** [comment]
Ah nice. I think we can probably use `Eval` here too, come to think of it?

**Comment by paf31** [comment]
Yes, I think so. We just need two actions to store and retrieve values along with names.

**Comment by hdgarrood** [comment]
Oh I guess we can't do it with the current `Eval`, because you can only obtain an `Eff eff Unit` with it right now. Perhaps we can make a minor modification to the class to allow supporting `<-` too.

**Comment by paf31** [comment]
I think you need something like

```purescript
class Eval comp result name | comp -> name, name -> result where
  eval :: comp -> Eff _ Unit
  evalAndStore :: comp -> Eff _ name
  retrieve :: name -> Eff _ result
```

**Comment by hdgarrood** [comment]
It's not clear to me that the storage and retrieval code should go in the `Eval` instance; wouldn't it make more sense to put that code in the psci backend?

I was imagining something more like this:

```purescript
class Eval a result | a -> result where
  eval :: a -> Eff _ result

evalPrint :: forall a result. (Show result, Eval a result) => a -> Eff _ Unit
```

**Comment by paf31** [comment]
This requires `result` to be serializable so that it can be passed to the Haskell process for storage.

**Comment by hdgarrood** [comment]
Ah, of course. Would it make sense to also have superclasses `EncodeJson result, DecodeJson result` then? Or do you prefer your original suggestion?

**Comment by paf31** [comment]
Well we can't depend on Argonaut here 😄 I would prefer to keep the values on the JS side.

**Comment by srghma** [comment]
it would be good not only support `<-` notation, [but also `<-` notation for `Aff` monad](https://github.com/purescript/purescript-psci-support/issues/16)
